api/rosetta/parse.go

```diff
@@ -0,0 +1,104 @@
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// ParseRequest implements the request schema for /construction/parse.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#request-4
+type ParseRequest struct {
+	NetworkID   identifier.Network `json:"network_identifier"`
+	Signed      bool               `json:"signed"`
+	Transaction string             `json:"transaction"`
+}
+
+// ParseResponse implements the response schema for /construction/parse.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#response-4
+type ParseResponse struct {
+	Operations []object.Operation   `json:"operations"`
+	SignerIDs  []identifier.Account `json:"account_identifier_signers,omitempty"`
+	Metadata   object.Metadata      `json:"metadata,omitempty"`
+}
+
+// Parse implements the /construction/parse endpoint of the Rosetta Construction API.
+// Parse endpoint parses both signed and unsigned transactions to understand the
+// transaction's intent. Endpoint returns the list of operations, any relevant metadata,
+// and, in the case of signed transaction, the list of signers.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructionparse
+func (c *Construction) Parse(ctx echo.Context) error {
+
+	var req ParseRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	if req.Transaction == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(txBodyEmpty))
+	}
+
+	var tx flow.Transaction
+	err = json.Unmarshal([]byte(req.Transaction), &tx)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(txBodyInvalid, withError(err)))
+	}
+
+	// If the transaction is signed, make sure that we have signatures provided.
+	if req.Signed && len(tx.EnvelopeSignatures) == 0 {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidFormat(txSignaturesEmpty))
+	}
+
+	// If the transaction is unsigned, verify that no signatures are present.
+	if !req.Signed && len(tx.EnvelopeSignatures) > 0 {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidFormat(txSignaturesFound))
+	}
+
+	// Parse the transaction and recreate the original list of operations, as well as all signers involved.
+	operations, signers, err := c.parser.ParseTransaction(&tx)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(txBodyInvalid, withError(err)))
```

Comment:
Errors returned by the function can be:
- invalid number of authorizers (want one)
- invalid number of script arguments (want two)
- errors in parsing arguments (one uint64 and one address)

I think all could fit okay-ish in the interpretation that the `transaction text is invalid`.

Since the input transaction text should be generated by us, I don't think any of those should happen in normal operation.

Reply:
Yes, I think that's just part of the security process Coinbase has implemented, to double-check they were fed the right information at every step. However, I would _still_ return different errors here. Just to be consistent and keep it as nice as possible.